File: server/lib/spy_hunt/gateway/gateway.ex
#
#  gateway.ex
#  server
#
#  Created by d-exclaimation on 17:02.
#

defmodule SpyHunt.Gateway do
  @moduledoc """
  Gateway GenServer, prior to entering the gateway all process must be taken as asynchronous
  """
  use GenServer

  alias SpyHunt.Socket.Client

  @type state :: %{
          lobbies: [any()]
        }

  @impl true
  @spec init(any()) :: {:ok, state()}
  def init(:ok) do
    {:ok, %{lobbies: []}}
  end

  @impl true
  @spec handle_call({:join, Client.t()}, GenServer.from(), state()) :: {:reply, %{}, state()}
  def handle_call({:join, %Client{} = c}, _from, %{lobbies: lobbies}) do
    case Enum.find_index(lobbies, fn lobby -> Enum.count(lobby) < 2 end) do
      nil ->
        new_lobbies = [[c] | lobbies]
        {:reply, %{}, %{lobbies: new_lobbies}}

      i ->
        new_lobbies =
          lobbies
          |> Enum.with_index()
          |> Enum.map(fn {lobby, j} ->
            if i == j, do: [c | lobby], else: lobby
          end)

        {:reply, %{}, %{lobbies: new_lobbies}}
    end
  end
end

File: server/src/main/scala/actor/Gateway.scala
//
//  Gateway.scala
//  spy-hunt
//
//  Created by d-exclaimation on 8:19 PM.
//

package actor

import actor.gateway._
import akka.actor.typed.scaladsl.ActorContext

/** Gateway GenServer, prior to entering the gateway all process must be taken as asynchronous */
object Gateway extends GenServer[Msg, State, Unit] {
  def init(params: Unit): Init =
    Init.Ok(
      State(lobbies = Seq.empty)
    )

  def cast(context: ActorContext[Msg], msg: Msg, state: State): Cast = msg match {
    case Msg.Join(client) =>
      val State(lobbies) = state

      lobbies.indexWhere(_.length < 2) match {
        case -1 =>
          Cast.NoReply(
            State(lobbies = lobbies :+ Seq(client))
          )

        case i =>
          val newLobbies = lobbies
            .zipWithIndex
            .map { case (lobby, j) =>
              if (i == j) lobby :+ client else lobby
            }
          Cast.NoReply(
            State(lobbies = newLobbies)
          )
      }

    case Msg.Test =>
      println(s"$state")
      Cast.NoReply(state)
  }

  def terminate(reason: String, state: State): Unit = ()
}
